design-decisions/2018-10-12-Condition-Split.md

```diff
@@ -0,0 +1,159 @@
+Splitting configurations upon new constraints on symbolic variables
+===================================================================
+
+Background
+----------
+
+When applying an axiom on a symbolic configuration new constraints may be
+introduced on symbolic variables, coming from side conditions, or directly
+from the unification process.
+
+### Example (from side conditions)
+
+Consider rewriting term `3 / A:Int` using
+
+`rule X:Int / Y:Int => X /Int Y requires Y =/=Int 0`
+which would be written as
+`Y =/= 0 = true ∧ X:Int / Y:Int => X /Int Y`
+
+Then `3 / A:Int ∧ (Y =/=Int 0 = true ∧ X:Int / Y:Int)` will simplify to
+```
+3 / A:Int ∧ (X:Int = 3 ∧ Y:Int = A:Int)
+          ∧ (A:Int =/=Int 0) = true
+```
+which is satisfiable, but not `T`.
+
+### Example (from unification)
+
+Consider rewriting term `if A:Int <Int 0 then -1 else 1` using
+
+`rule if true then E1:Exp else E2:Exp => E1`
+
+Then `if A:Int <Int 0 then -1 else 1 ∧ if true then E1:Exp else E2:Exp` will
+simplify to
+```
+if A:Int <Int 0 then -1 else 1 ∧ (E1:Exp = -1 ∧ E2:Exp = 1)
+                               ∧ A:Int <Int 0 = true
+```
+which is satisfiable, but not `T`.
+
+Problem
+-------
+
+Consider the following "one-path" reachability problem
+
+```
+∀ X . φ(X) -> ◇ ∃ Y . ψ(X, Y)
+```
+where `X` and `Y` are sets of variables.
+
+### Reduction to proving a ○ implication
+
+By unrolling ◇ we obtain:
+```
+∀ X . φ(X) ->  ∃ Y . ψ(X, Y) ∨ ○ ◇ ∃ Y . ψ(X, Y)
+```
+
+We can now move `∃ Y . ψ(X, Y)` to the right of the implication,
+and (assuming law of excluded middle) we obtain the equivalent:
+```
+(∀ X . (φ(X) ∧ ¬∃ Y . ψ(X, Y)) -> ○ ◇ ∃ Y . ψ(X, Y))
+```
+
+#### Basic Symbolic Execution
+
+**Input:** `φ(X)`, `∀ Y .α(Y) -> ○ β(Y)`
+
+**Output:** a pattern `φ'(X)` and satisfying the same assumptions we put on
```

Comment:
What assumption on φ? No assumptions have been given about the form of any formulas so far?